src/System.IO.Ports/src/System/IO/Ports/SerialStream.Unix.cs

```diff
@@ -441,55 +522,49 @@ internal unsafe void Write(byte[] array, int offset, int count, int timeout)
                 throw new ArgumentOutOfRangeException(nameof(offset), SR.ArgumentOutOfRange_NeedPosNum);
             if (count < 0)
                 throw new ArgumentOutOfRangeException(nameof(count), SR.ArgumentOutOfRange_NeedPosNum);
-            if (count == 0) return; // no need to expend overhead in creating asyncResult, etc.
             if (array.Length - offset < count)
                 throw new ArgumentException(SR.Argument_InvalidOffLen);
 
             // check for open handle, though the port is always supposed to be open
-            if (_handle == null) throw new ObjectDisposedException(SR.Port_not_open);
-
-            int numBytes = 0;
+            if (_handle == null)
+                throw new ObjectDisposedException(SR.Port_not_open);
 
-            while (count > 0)
-            {
-                if (timeout > 0)
-                {
-                    Interop.Sys.PollEvents events = Interop.Sys.PollEvents.POLLNONE;
-                    Interop.Sys.Poll(_handle, Interop.Sys.PollEvents.POLLOUT | Interop.Sys.PollEvents.POLLERR, timeout,out events);
-
-                    if ((events & (Interop.Sys.PollEvents.POLLERR | Interop.Sys.PollEvents.POLLNVAL)) != 0)
-                    {
-                        throw new IOException();
-                    }
+            Memory<byte> buffer = new Memory<byte>(array, offset, count);
+            return BeginWriteCore(SerialPort.InfiniteTimeout, buffer, userCallback, stateObject);
+        }
 
-                    if ( (events & Interop.Sys.PollEvents.POLLOUT) == 0)
-                    {
-                        throw new TimeoutException(SR.Write_timed_out);
-                    }
-                }
+        private IAsyncResult BeginWriteCore(int timeout, Memory<byte> buffer, AsyncCallback userCallback, object stateObject)
+        {
+            SerialStreamAsyncResult result = new SerialStreamAsyncResult(
+                GetCancellationTokenFromTimeout(timeout),
+                buffer);
+            _writeQueue.Enqueue(result);
 
-                fixed (byte* bufPtr = array)
-                {
-                    numBytes = Interop.Sys.Write(_handle, bufPtr + offset, count);
-                }
+            return TaskToApm.Begin((Task)result.UnderlyingTask, userCallback, stateObject);
+        }
```

Comment:
The Begin/End methods can be implemented in terms of the XxAsync methods, rather than duplicating the logic.

Reply:
I mean just:
```C#
return TaskToApm.Begin(WriteAsync(…), userCallback, stateObject);
```
rather than effectively duplicating the implementation of WriteAsync here.